File: Clients/Node/Core/src/services/TasksService.ts
import type { FetchBasedClient } from '../client/FetchBasedClient';
import { HttpMethod } from '../client/HttpMethod';
import type { RequestOptions } from '../client/types';
import { ConduitError } from '../utils/errors';
import { API_ENDPOINTS, TASK_STATUS } from '../constants';

/**
 * Service for general task management operations using the Conduit Core API
 */
export class TasksService {

  constructor(private readonly client: FetchBasedClient) {}

  /**
   * Gets the status of any task by its ID
   */
  async getTaskStatus(
    taskId: string,
    options?: RequestOptions
  ): Promise<TaskStatusResponse> {
    try {
      if (!taskId || taskId.trim().length === 0) {
        throw new Error('Task ID is required');
      }

      const response = await this.client['request']<TaskStatusResponse>(
        API_ENDPOINTS.V1.TASKS.BY_ID(taskId),
        {
          method: HttpMethod.GET,
          ...options
        }
      );

      return response;
    } catch (error) {
      if (error instanceof ConduitError) {
        throw error;
      }
      throw new ConduitError(
        `Failed to get task status: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Cancels a pending or running task
   */
  async cancelTask(
    taskId: string,
    options?: RequestOptions
  ): Promise<void> {
    try {
      if (!taskId || taskId.trim().length === 0) {
        throw new Error('Task ID is required');
      }

      await this.client['request']<void>(
        API_ENDPOINTS.V1.TASKS.CANCEL(taskId),
        {
          method: HttpMethod.POST,
          body: {},
          ...options
        }
      );
    } catch (error) {
      if (error instanceof ConduitError) {
        throw error;
      }
      throw new ConduitError(
        `Failed to cancel task: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Polls a task until completion or timeout
   */
  async pollTaskUntilCompletion<T = any>(
    taskId: string,
    pollingOptions?: TaskPollingOptions,
    options?: RequestOptions
  ): Promise<T> {
    const opts: Required<TaskPollingOptions> = {
      intervalMs: pollingOptions?.intervalMs ?? TaskDefaults.POLLING_INTERVAL_MS,
      timeoutMs: pollingOptions?.timeoutMs ?? TaskDefaults.POLLING_TIMEOUT_MS,
      useExponentialBackoff: pollingOptions?.useExponentialBackoff ?? true,
      maxIntervalMs: pollingOptions?.maxIntervalMs ?? TaskDefaults.MAX_POLLING_INTERVAL_MS,
    };

    if (!taskId || taskId.trim().length === 0) {
      throw new Error('Task ID is required');
    }

    const startTime = Date.now();
    let currentInterval = opts.intervalMs;

    while (true) {
      // Check timeout
      if (Date.now() - startTime > opts.timeoutMs) {
        throw new ConduitError(
          `Task polling timed out after ${opts.timeoutMs}ms`
        );
      }

      const status = await this.getTaskStatus(taskId, options);

      switch (status.status?.toLowerCase()) {
        case TASK_STATUS.COMPLETED:
          if (!status.result) {
            throw new ConduitError(
              'Task completed but no result was provided'
            );
          }
          return status.result as T;

        case TASK_STATUS.FAILED:
          throw new ConduitError(
            `Task failed: ${status.error || 'Unknown error'}`
          );

        case TASK_STATUS.CANCELLED:
          throw new ConduitError('Task was cancelled');

        case TASK_STATUS.TIMEDOUT:
          throw new ConduitError('Task timed out');

        case TASK_STATUS.PENDING:
        case TASK_STATUS.RUNNING:
          // Continue polling
          break;

        default:
          throw new ConduitError(
            `Unknown task status: ${status.status}`
          );
      }

      // Wait before next poll
      await this.sleep(currentInterval);

      // Apply exponential backoff if enabled
      if (opts.useExponentialBackoff) {
        currentInterval = Math.min(currentInterval * 2, opts.maxIntervalMs);
      }
    }
  }

  /**
   * Requests cleanup of old completed tasks (admin operation)
   */
  async cleanupOldTasks(
    olderThanHours: number = 24,
    options?: RequestOptions
  ): Promise<number> {
    try {
      const request = { older_than_hours: olderThanHours };

      const response = await this.client['request']<CleanupTasksResponse>(
        API_ENDPOINTS.V1.TASKS.CLEANUP,
        {
          method: HttpMethod.POST,
          body: request,
          ...options
        }
      );

      return response.tasks_removed;
    } catch (error) {
      if (error instanceof ConduitError) {
        throw error;
      }
      throw new ConduitError(
        `Failed to cleanup old tasks: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Helper method to sleep for a specified duration
   */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

/**
 * Response from a general task status request
 */
export interface TaskStatusResponse {
  /** The unique task identifier */
  task_id: string;

  /** The current status of the task */
  status?: string;

  /** The task type */
  task_type?: string;

  /** The progress percentage (0-100) */
  progress: number;

  /** An optional progress message */
  message?: string;

  /** When the task was created */
  created_at: string;

  /** When the task was last updated */
  updated_at: string;

  /** The task result, available when status is completed */
  result?: any;

  /** Error information if the task failed */
  error?: string;
}

/**
 * Options for polling task status
 */
export interface TaskPollingOptions {
  /** The polling interval in milliseconds */
  intervalMs?: number;

  /** The maximum polling timeout in milliseconds */
  timeoutMs?: number;

  /** Whether to use exponential backoff for polling intervals */
  useExponentialBackoff?: boolean;

  /** The maximum interval between polls in milliseconds when using exponential backoff */
  maxIntervalMs?: number;
}

/**
 * Response from a cleanup tasks request
 */
export interface CleanupTasksResponse {
  /** The number of tasks that were removed */
  tasks_removed: number;
}

/**
 * Default values for task operations
 */
export const TaskDefaults = {
  /** Default polling interval in milliseconds */
  POLLING_INTERVAL_MS: 2000,

  /** Default polling timeout in milliseconds */
  POLLING_TIMEOUT_MS: 600000, // 10 minutes

  /** Default maximum polling interval in milliseconds */
  MAX_POLLING_INTERVAL_MS: 30000, // 30 seconds
} as const;

/**
 * Helper functions for task management
 */
export const TaskHelpers = {
  /**
   * Creates polling options with sensible defaults
   */
  createPollingOptions(
    options?: Partial<TaskPollingOptions>
  ): TaskPollingOptions {
    return {
      intervalMs: options?.intervalMs ?? TaskDefaults.POLLING_INTERVAL_MS,
      timeoutMs: options?.timeoutMs ?? TaskDefaults.POLLING_TIMEOUT_MS,
      useExponentialBackoff: options?.useExponentialBackoff ?? true,
      maxIntervalMs: options?.maxIntervalMs ?? TaskDefaults.MAX_POLLING_INTERVAL_MS,
    };
  },
};

File: archive/Clients.DotNet/ConduitLLM.CoreClient/src/Services/TasksService.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using ConduitLLM.CoreClient.Client;
using ConduitLLM.CoreClient.Constants;
using ConduitLLM.CoreClient.Exceptions;
using ConduitLLM.CoreClient.Models;
using ConduitLLM.CoreClient.Utils;

namespace ConduitLLM.CoreClient.Services
{
    /// <summary>
    /// Service for general task management operations using the Conduit Core API.
    /// </summary>
    public class TasksService
    {
        private readonly BaseClient _client;
        private readonly ILogger<TasksService>? _logger;
        private const string TasksEndpoint = ApiEndpoints.V1.Tasks.Base;

        /// <summary>
        /// Initializes a new instance of the <see cref="TasksService"/> class.
        /// </summary>
        /// <param name="client">The base client for making API requests.</param>
        /// <param name="logger">Optional logger for debugging and monitoring.</param>
        public TasksService(BaseClient client, ILogger<TasksService>? logger = null)
        {
            _client = client ?? throw new ArgumentNullException(nameof(client));
            _logger = logger;
        }

        /// <summary>
        /// Gets the status of any task by its ID.
        /// </summary>
        /// <param name="taskId">The task identifier.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The current task status and result if completed.</returns>
        /// <exception cref="ValidationException">Thrown when the task ID is invalid.</exception>
        /// <exception cref="ConduitCoreException">Thrown when the API request fails.</exception>
        public async Task<TaskStatusResponse> GetTaskStatusAsync(
            string taskId,
            CancellationToken cancellationToken = default)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(taskId))
                    throw new ValidationException("Task ID is required", "taskId");

                var endpoint = $"{TasksEndpoint}/{Uri.EscapeDataString(taskId)}";
                
                var response = await _client.GetForServiceAsync<TaskStatusResponse>(
                    endpoint,
                    cancellationToken);

                _logger?.LogDebug("Retrieved status for task {TaskId}: {Status}", 
                    taskId, response.Status);
                
                return response;
            }
            catch (Exception ex) when (!(ex is ConduitCoreException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        /// <summary>
        /// Cancels a pending or running task.
        /// </summary>
        /// <param name="taskId">The task identifier.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <exception cref="ValidationException">Thrown when the task ID is invalid.</exception>
        /// <exception cref="ConduitCoreException">Thrown when the API request fails.</exception>
        public async Task CancelTaskAsync(
            string taskId,
            CancellationToken cancellationToken = default)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(taskId))
                    throw new ValidationException("Task ID is required", "taskId");

                var endpoint = $"{TasksEndpoint}/{Uri.EscapeDataString(taskId)}/cancel";
                
                await _client.PostForServiceAsync<object>(endpoint, new { }, cancellationToken);
                
                _logger?.LogDebug("Cancelled task {TaskId}", taskId);
            }
            catch (Exception ex) when (!(ex is ConduitCoreException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }

        /// <summary>
        /// Polls a task until completion or timeout.
        /// </summary>
        /// <param name="taskId">The task identifier.</param>
        /// <param name="options">Polling options.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>The final task result when completed.</returns>
        /// <exception cref="ValidationException">Thrown when parameters are invalid.</exception>
        /// <exception cref="TimeoutException">Thrown when polling times out.</exception>
        /// <exception cref="ConduitCoreException">Thrown when the API request fails or task fails.</exception>
        public async Task<T> PollTaskUntilCompletionAsync<T>(
            string taskId,
            TaskPollingOptions? options = null,
            CancellationToken cancellationToken = default) where T : class
        {
            options ??= new TaskPollingOptions();
            
            if (string.IsNullOrWhiteSpace(taskId))
                throw new ValidationException("Task ID is required", "taskId");

            var startTime = DateTime.UtcNow;
            var currentInterval = options.IntervalMs;
            
            _logger?.LogDebug("Starting to poll task {TaskId} with interval {IntervalMs}ms, timeout {TimeoutMs}ms", 
                taskId, options.IntervalMs, options.TimeoutMs);

            while (true)
            {
                cancellationToken.ThrowIfCancellationRequested();

                // Check timeout
                if ((DateTime.UtcNow - startTime).TotalMilliseconds > options.TimeoutMs)
                {
                    throw new TimeoutException($"Task polling timed out after {options.TimeoutMs}ms");
                }

                var status = await GetTaskStatusAsync(taskId, cancellationToken);

                switch (status.Status?.ToLowerInvariant())
                {
                    case "completed":
                        if (status.Result == null)
                            throw new ConduitCoreException("Task completed but no result was provided", null, null, null, null);
                        
                        _logger?.LogDebug("Task {TaskId} completed successfully", taskId);
                        return status.Result as T ?? throw new ConduitCoreException("Task result is not of expected type", null, null, null, null);

                    case "failed":
                        throw new ConduitCoreException($"Task failed: {status.Error ?? "Unknown error"}", null, null, null, null);

                    case "cancelled":
                        throw new ConduitCoreException("Task was cancelled", null, null, null, null);

                    case "timedout":
                        throw new ConduitCoreException("Task timed out", null, null, null, null);

                    case "pending":
                    case "running":
                        // Continue polling
                        break;

                    default:
                        throw new ConduitCoreException($"Unknown task status: {status.Status}", null, null, null, null);
                }

                // Wait before next poll
                await Task.Delay(currentInterval, cancellationToken);

                // Apply exponential backoff if enabled
                if (options.UseExponentialBackoff)
                {
                    currentInterval = Math.Min(currentInterval * 2, options.MaxIntervalMs);
                }
            }
        }

        /// <summary>
        /// Requests cleanup of old completed tasks (admin operation).
        /// </summary>
        /// <param name="olderThanHours">Remove tasks older than this number of hours.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <returns>Number of tasks cleaned up.</returns>
        /// <exception cref="ConduitCoreException">Thrown when the API request fails.</exception>
        public async Task<int> CleanupOldTasksAsync(
            int olderThanHours = 24,
            CancellationToken cancellationToken = default)
        {
            try
            {
                var endpoint = $"{TasksEndpoint}/cleanup";
                var request = new { older_than_hours = olderThanHours };
                
                var response = await _client.PostForServiceAsync<CleanupTasksResponse>(
                    endpoint, 
                    request, 
                    cancellationToken);
                
                _logger?.LogDebug("Cleaned up {Count} old tasks", response.TasksRemoved);
                return response.TasksRemoved;
            }
            catch (Exception ex) when (!(ex is ConduitCoreException))
            {
                ErrorHandler.HandleException(ex);
                throw;
            }
        }
    }

    /// <summary>
    /// Represents the response from a general task status request.
    /// </summary>
    public class TaskStatusResponse
    {
        /// <summary>
        /// Gets or sets the unique task identifier.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("task_id")]
        public string TaskId { get; set; } = string.Empty;

        /// <summary>
        /// Gets or sets the current status of the task.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("status")]
        public string? Status { get; set; }

        /// <summary>
        /// Gets or sets the task type.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("task_type")]
        public string? TaskType { get; set; }

        /// <summary>
        /// Gets or sets the progress percentage (0-100).
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("progress")]
        public int Progress { get; set; }

        /// <summary>
        /// Gets or sets an optional progress message.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("message")]
        public string? Message { get; set; }

        /// <summary>
        /// Gets or sets when the task was created.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("created_at")]
        public DateTime CreatedAt { get; set; }

        /// <summary>
        /// Gets or sets when the task was last updated.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("updated_at")]
        public DateTime UpdatedAt { get; set; }

        /// <summary>
        /// Gets or sets the task result, available when status is completed.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("result")]
        public object? Result { get; set; }

        /// <summary>
        /// Gets or sets error information if the task failed.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("error")]
        public string? Error { get; set; }
    }

    /// <summary>
    /// Represents options for polling task status.
    /// </summary>
    public class TaskPollingOptions
    {
        /// <summary>
        /// Gets or sets the polling interval in milliseconds.
        /// </summary>
        public int IntervalMs { get; set; } = 2000;

        /// <summary>
        /// Gets or sets the maximum polling timeout in milliseconds.
        /// </summary>
        public int TimeoutMs { get; set; } = 600000; // 10 minutes

        /// <summary>
        /// Gets or sets whether to use exponential backoff for polling intervals.
        /// </summary>
        public bool UseExponentialBackoff { get; set; } = true;

        /// <summary>
        /// Gets or sets the maximum interval between polls in milliseconds when using exponential backoff.
        /// </summary>
        public int MaxIntervalMs { get; set; } = 30000; // 30 seconds
    }

    /// <summary>
    /// Represents the response from a cleanup tasks request.
    /// </summary>
    public class CleanupTasksResponse
    {
        /// <summary>
        /// Gets or sets the number of tasks that were removed.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("tasks_removed")]
        public int TasksRemoved { get; set; }
    }
}